transaction/DirectedGraph.hpp

```diff
@@ -0,0 +1,206 @@
+  }
+
+  /**
+   * @brief Adds a new node to the graph with the given transaction id.
+   *        It checks whether the transaction id is valid or not.
+   * @warning Pointer ownership will pass to the graph, therefore it
+   *          should not be deleted.
+   *
+   * @param data Pointer to the transaction id that will be contained
+   *        in the node.
+   * @return Id of the newly created node.
+   **/
+  inline node_id addNodeCheckExists(transaction_id *data) {
+    for (std::vector<DirectedGraphNode>::const_iterator
+           it = nodes_.cbegin(), end = nodes_.cend();
+         it != end;
+         ++it) {
+      CHECK(*data != it->getData());
+    }
+    nodes_.emplace_back(data);
+    return nodes_.size() - 1;
+  }
+
+  /**
+   * @brief Adds an edge between nodes. It does not check whether the
+   *        parameters are valid node ids.
+   * @warning Does not check arguments are legit. It may cause
+   *          out of range errors.
+   *
+   * @param fromNode The node that edge is orginated.
+   * @param toNode The node that edge is ended.
+   **/
+  inline void addEdgeUnchecked(node_id from_node, node_id to_node) {
+    nodes_[from_node].addOutgoingEdge(to_node);
+  }
+
+  /**
+   * @brief Adds an edge between nodes. It checks whether the
+   *        parameters are valid node ids.
+   *
+   * @param fromNode The node that edge is orginated.
+   * @param toNode The node that edge is ended.
+   **/
+  inline void addEdgeCheckExists(node_id from_node, node_id to_node) {
+    CHECK(from_node < getNumNodes() && to_node < getNumNodes());
+    nodes_[from_node].addOutgoingEdge(to_node);
+  }
+
+  /**
+   * @brief Check whether there is a directed edge.
+   * @warning When parameters are not valid ids, it may cause
+   *          out of range errors.
+   *
+   * @param fromNode Id of the node that edge is originated from.
+   * @param toNode Id of the node that edge is ended.
+   * @return True if there is an edge, false otherwise.
+   **/
+  inline bool hasEdge(node_id from_node, node_id to_node) const {
+    DCHECK(from_node < getNumNodes() && to_node < getNumNodes());
+    return nodes_[from_node].hasOutgoingEdge(to_node);
+  }
+
+  /**
+   * @brief Get data (transaction id) contained in the node.
+   * @warning Does not check index validity.
+   *
+   * @param node Id of the node that the data is got from.
+   * @return Id of the transaction that this node contains.
+   **/
+  inline transaction_id getDataFromNode(node_id node) const {
+    return nodes_[node].getData();
+  }
+
+  /**
+   * @brief Calculate how many nodes the graph has.
+   *
+   * @return The number of nodes the graph has.
+   **/
+  inline std::size_t getNumNodes() const {
+    return nodes_.size();
+  }
+
+  /**
+   * @brief Gives the node ids that this node has edges to.
+   *
+   * @param id Id of the corresponding node.
+   * @return Vector of node ids that id has edges to.
+   **/
+  inline std::vector<node_id> getAdjacentNodes(node_id id) const {
```

Comment:
Harshad, I thought the alternatives. Storing the data in vector, and checking the uniqueness in each `addEdge` operation, is more expensive since the number of `addEdge` calls is more than the number of `getAdjacentNodes` calls. I think, we should keep it this way, for now.